util/flocks/src/frwlock/read_guard.rs

```diff
@@ -0,0 +1,25 @@
+use std::ops::Deref;
+use rustix::{
+    fs::{flock, FlockOperation},
+    fd::AsFd
+};
+
+pub struct FrwLockReadGuard<T: AsFd> {
+    pub(crate) data: *const T
+}
```

Comment:
What's the guarantee that the raw pointer will stay valid for the lifetime of the guard object?

Reply:
Very important catch. I had rewritten to `OwnedGuards` because the lifetimes were becoming quite difficult to manage: https://github.com/tokio-rs/tokio/blob/06582776a564c88a7b4c6f9e3d7c0ebd0ef3f34b/tokio/src/sync/rwlock.rs#L905. But, I need to change the receivers to Arc<Self> and probably rename these. 


Otherwise, yes this needs...
```
pub(crate) data: *mut T,
pub(crate) marker: PhantomData<&'a mut T>
```

Reply:
The normal Rust way to avoid the unsafety for temporary RAII objects is to use lifetimes, and `rustix` provides some additional help:

```rust
use rustix::fd::BorrowedFd;

pub struct FrwLockReadGuard<T, 'a> {
    fd: BorrowedFd<'a>
}
```

`BorrowedFd` satisfies `Send + Sync`. I think the lifetime bound would let you code async blocks safely without allowing any locks to "escape".

Reply:
A phantom marker is not enough, you have to guarantee the validity of the pointee object. An `Arc` over the file object would guarantee that. But I expect owned locks to be bug-prone in general.